### PR TITLE
Laboratory work #3, Ivan Shmakov - 23FPL1

### PR DESCRIPTION
**Pull request by CloudWhisp**

### CODE REVIEW

File: lab_3_ann_retriever/main.py [hunk 1]
```diff
@@ -104,6 +114,14 @@ def tokenize(self, text: str) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(text, str) or not isinstance(self, Tokenizer):
+            return None
+        for sign in text.lower():
+            if not sign.isalpha():
+                text = text.lower().replace(sign, ' ')
+        text = text.replace('  ', ' ')
+        text = self._remove_stop_words(text.split())
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:123]: do not create the variable

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -130,7 +159,19 @@ def _remove_stop_words(self, tokens: list[str]) -> list[str] | None:
...
+            return None
+        if tokens == [] or self._stop_words == []:
+            return None
+        for data in tokens:
+            if not isinstance(data, str):
+                return None
+        for data in self._stop_words:
+            if not isinstance(data, str):
+                return None
+        tokens_cleared = [elem for elem in tokens if elem not in self._stop_words]
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:173]: .

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -170,7 +232,10 @@ def vectorize(self, tokenized_document: list[str]) -> Vector | None:
 
         In case of corrupt input arguments, None is returned.
         """
-
+        if not isinstance(tokenized_document, list) or len(tokenized_document) == 0:
+            return None
+        vector = self._calculate_tf_idf(tokenized_document)
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:237]: .

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -321,6 +444,17 @@ def _calculate_knn(
...
+        if (not isinstance(query_vector, tuple) or not isinstance(document_vectors, list)
+            or not isinstance(n_neighbours, int) or len(document_vectors) == 0):
+            return None
+        docs = []
+        for ind, doc in enumerate(document_vectors):
+            if not isinstance(doc, tuple) or len(doc) == 0:
+                return None
+            distance = calculate_distance(query_vector, doc)
+            docs.append((int(ind), distance))
+        docs = sorted(docs, key= lambda x:x[1])
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:456]: .

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -334,7 +468,11 @@ def _index_document(self, document: str) -> Vector | None:
 
         In case of corrupt input arguments, None is returned.
         """
-
+        if not isinstance(document, str) or len(document) == 0:
+            return None
+        tokenized_text = self._tokenizer.tokenize(document)
+        vector = self._vectorizer.vectorize(tokenized_text)
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:474]: .

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -381,7 +519,10 @@ def __init__(
             left_node (NodeLike | None): Left node
             right_node (NodeLike | None): Right node
         """
-
+        self.vector = vector
+        self.left_node = left_node
+        self.right_node = right_node
+        self.payload = payload
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:525]: add empty line

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -413,7 +554,7 @@ def __init__(self) -> None:
         """
         Initialize an instance of the KDTree class.
         """
-
+        self._root = None
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:557]: add empty line

File: lab_3_ann_retriever/start.py [hunk 8]
```diff
@@ -30,7 +30,20 @@ def main() -> None:
...
     with open("assets/secrets/secret_1.txt", "r", encoding="utf-8") as text_file:
         text = text_file.read()
-    result = None
+    documents, stopwords = open_files()
+    query = 'Нижний Новгород'
+    tokenizer = Tokenizer(stopwords)
+    docs_tokenized = tokenizer.tokenize_documents(documents)
+    vectorizer = Vectorizer(docs_tokenized)
+
+    #knn_retriever = BasicSearchEngine(vectorizer=vectorizer, tokenizer=tokenizer)
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/start.py:39]: remove

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (CloudWhisp) [reply lab_3_ann_retriever/main.py:237]: Done

File: lab_3_ann_retriever/main.py (hunk 5 above)
Comment (CloudWhisp) [reply lab_3_ann_retriever/main.py:474]: Done

File: lab_3_ann_retriever/main.py (hunk 7 above)
Comment (CloudWhisp) [reply lab_3_ann_retriever/main.py:557]: Fixed

File: lab_3_ann_retriever/start.py (hunk 8 above)
Comment (CloudWhisp) [reply lab_3_ann_retriever/start.py:43]: Done

**Pull request closed without merging**